### ISSUE TITLE
Unable to override local python interpreter with "connection: local" unless on localhost

### ISSUE THREAD

**Issue by copumpkin** [ansible/ansible#13421]
### Issue Type:

Bug report
### Ansible Version:

```
ansible 1.9.4
  configured module search path = None
```
### Ansible Configuration:

N/A, minimized below.
### Environment:

N/A, minimized below.
### Summary:

It doesn't seem possible to override `ansible_python_interpreter` when using `connection: local`. Minimized example below. cc @bcoca whom I spoke with about this on IRC.
### Steps To Reproduce:

test.yml:

``` yaml
- hosts: localhost
  tasks:
    - debug:
        msg: wow!

- hosts: other
  connection: local
  tasks:
    - debug:
        msg: wow such local!
```

Now run `ansible-playbook -vvvv -i "other," test.yml`
### Expected Results:

Both tasks run with my custom `python` interpreter.
### Actual Results:

Only the first task (running against `localhost`) uses my custom `python` interpreter. The second one, which uses `connection: local` in the context of a different host, uses `/usr/bin/python`, presumably from the shebang of the module it's running.

Here's my actual log:

```
$ ansible-playbook -vvvv -i inventory test.yml 

PLAY [localhost] ************************************************************** 

GATHERING FACTS *************************************************************** 
<localhost> REMOTE_MODULE setup
<localhost> EXEC ['/bin/sh', '-c', 'mkdir -p $HOME/.ansible/tmp/ansible-tmp-1449184980.94-175599515598394 && chmod a+rx $HOME/.ansible/tmp/ansible-tmp-1449184980.94-175599515598394 && echo $HOME/.ansible/tmp/ansible-tmp-1449184980.94-175599515598394']
<localhost> PUT /var/folders/f2/18n30h057611nz0cy1ng9bhc0000gn/T/tmpoj1H3_ TO /Users/copumpkin/.ansible/tmp/ansible-tmp-1449184980.94-175599515598394/setup
<localhost> EXEC ['/bin/sh', '-c', u'LANG=en_US.UTF-8 LC_CTYPE=en_US.UTF-8 /nix/store/5gzrrcw9g3477d354l5df3w6cqyjvimc-python-2.7.10/bin/python2.7 /Users/copumpkin/.ansible/tmp/ansible-tmp-1449184980.94-175599515598394/setup; rm -rf /Users/copumpkin/.ansible/tmp/ansible-tmp-1449184980.94-175599515598394/ >/dev/null 2>&1']
ok: [localhost]

TASK: [debug ] **************************************************************** 
ok: [localhost] => {
    "msg": "wow!"
}

PLAY [other] ****************************************************************** 

GATHERING FACTS *************************************************************** 
<other> REMOTE_MODULE setup
<other> EXEC ['/bin/sh', '-c', 'mkdir -p $HOME/.ansible/tmp/ansible-tmp-1449184981.75-161849069102207 && chmod a+rx $HOME/.ansible/tmp/ansible-tmp-1449184981.75-161849069102207 && echo $HOME/.ansible/tmp/ansible-tmp-1449184981.75-161849069102207']
<other> PUT /var/folders/f2/18n30h057611nz0cy1ng9bhc0000gn/T/tmpXztwbB TO /Users/copumpkin/.ansible/tmp/ansible-tmp-1449184981.75-161849069102207/setup
<other> EXEC ['/bin/sh', '-c', u'LANG=en_US.UTF-8 LC_CTYPE=en_US.UTF-8 /usr/bin/python /Users/copumpkin/.ansible/tmp/ansible-tmp-1449184981.75-161849069102207/setup; rm -rf /Users/copumpkin/.ansible/tmp/ansible-tmp-1449184981.75-161849069102207/ >/dev/null 2>&1']
ok: [other]

TASK: [debug ] **************************************************************** 
ok: [other] => {
    "msg": "wow such local!"
}

PLAY RECAP ******************************************************************** 
localhost                  : ok=2    changed=0    unreachable=0    failed=0   
other                      : ok=2    changed=0    unreachable=0    failed=0   
```

Note the `/nix/store/...` python interpreter in the first task and `/usr/bin/python` in the second. The first one is also the value of `sys.executable` within `ansible-playbook`.

**Comment by copumpkin** [comment]
TBC, I'm relying on `localhost` being specified implicitly when I omit it from the inventory: https://github.com/ansible/ansible/blob/v1.9.4-1/lib/ansible/inventory/__init__.py#L319-L329

**Comment by copumpkin** [comment]
I just tried the same exact setup on v2.0.0-0.6.rc1 and got similar output. Perhaps it's intended behavior? In which case, I'm wondering how to override the python interpreter properly on my local connection.

```
$ ansible-playbook -vvvv -i inventory test.yml 
No config file found; using defaults
2 plays in test.yml
Loaded callback default of type stdout, v2.0

PLAY ***************************************************************************

TASK [setup] *******************************************************************
ESTABLISH LOCAL CONNECTION FOR USER: copumpkin
127.0.0.1 EXEC (umask 22 && mkdir -p "`echo $HOME/.ansible/tmp/ansible-tmp-1449197501.52-231707996663050`" && echo "`echo $HOME/.ansible/tmp/ansible-tmp-1449197501.52-231707996663050`")
127.0.0.1 PUT /var/folders/f2/18n30h057611nz0cy1ng9bhc0000gn/T/tmpR03Txk TO /Users/copumpkin/.ansible/tmp/ansible-tmp-1449197501.52-231707996663050/setup
127.0.0.1 EXEC LANG=en_US.UTF-8 LC_ALL=en_US.UTF-8 LC_MESSAGES=en_US.UTF-8 /nix/store/prc9bcrfz1ylzhlrrdgdmhbxdfmjf05x-python-2.7.10/bin/python2.7 /Users/copumpkin/.ansible/tmp/ansible-tmp-1449197501.52-231707996663050/setup; rm -rf "/Users/copumpkin/.ansible/tmp/ansible-tmp-1449197501.52-231707996663050/" > /dev/null 2>&1
ok: [localhost]

TASK [debug msg=wow!] **********************************************************
task path: /Users/copumpkin/Sandbox/ansible-repro/test.yml:3
ok: [localhost] => {
    "changed": false, 
    "invocation": {
        "module_args": {
            "msg": "wow!"
        }, 
        "module_name": "debug"
    }, 
    "msg": "wow!"
}

PLAY ***************************************************************************

TASK [setup] *******************************************************************
ESTABLISH LOCAL CONNECTION FOR USER: copumpkin
other EXEC (umask 22 && mkdir -p "`echo $HOME/.ansible/tmp/ansible-tmp-1449197502.33-180008873006103`" && echo "`echo $HOME/.ansible/tmp/ansible-tmp-1449197502.33-180008873006103`")
other PUT /var/folders/f2/18n30h057611nz0cy1ng9bhc0000gn/T/tmpOYFw9N TO /Users/copumpkin/.ansible/tmp/ansible-tmp-1449197502.33-180008873006103/setup
other EXEC LANG=en_US.UTF-8 LC_ALL=en_US.UTF-8 LC_MESSAGES=en_US.UTF-8 /usr/bin/python /Users/copumpkin/.ansible/tmp/ansible-tmp-1449197502.33-180008873006103/setup; rm -rf "/Users/copumpkin/.ansible/tmp/ansible-tmp-1449197502.33-180008873006103/" > /dev/null 2>&1
ok: [other]

TASK [debug msg=wow such local!] ***********************************************
task path: /Users/copumpkin/Sandbox/ansible-repro/test.yml:9
ok: [other] => {
    "changed": false, 
    "invocation": {
        "module_args": {
            "msg": "wow such local!"
        }, 
        "module_name": "debug"
    }, 
    "msg": "wow such local!"
}

PLAY RECAP *********************************************************************
localhost                  : ok=2    changed=0    unreachable=0    failed=0   
other                      : ok=2    changed=0    unreachable=0    failed=0   
```

Note that the two tasks have different python interpreters.

**Comment by copumpkin** [comment]
And same thing on devel at a1f516824ee2160121437edf6939ab2145972739. I must be doing something wrong.

```
$ ansible-playbook -vvvv -i inventory test.yml 
No config file found; using defaults
Loaded callback default of type stdout, v2.0
2 plays in test.yml

PLAY ***************************************************************************

TASK [setup] *******************************************************************
ESTABLISH LOCAL CONNECTION FOR USER: copumpkin
127.0.0.1 EXEC ( umask 22 && mkdir -p "$( echo $HOME/.ansible/tmp/ansible-tmp-1449199217.85-216850251156072 )" && echo "$( echo $HOME/.ansible/tmp/ansible-tmp-1449199217.85-216850251156072 )" )
127.0.0.1 PUT /var/folders/f2/18n30h057611nz0cy1ng9bhc0000gn/T/tmpilkScb TO /Users/copumpkin/.ansible/tmp/ansible-tmp-1449199217.85-216850251156072/setup
127.0.0.1 EXEC LANG=en_US.UTF-8 LC_ALL=en_US.UTF-8 LC_MESSAGES=en_US.UTF-8 /nix/store/prc9bcrfz1ylzhlrrdgdmhbxdfmjf05x-python-2.7.10/bin/python2.7 /Users/copumpkin/.ansible/tmp/ansible-tmp-1449199217.85-216850251156072/setup; rm -rf "/Users/copumpkin/.ansible/tmp/ansible-tmp-1449199217.85-216850251156072/" > /dev/null 2>&1
ok: [localhost]

TASK [debug msg=wow!] **********************************************************
task path: /Users/copumpkin/Sandbox/ansible-repro/test.yml:3
ok: [localhost] => {
    "changed": false, 
    "invocation": {
        "module_args": {
            "msg": "wow!"
        }, 
        "module_name": "debug"
    }, 
    "msg": "wow!"
}

PLAY ***************************************************************************

TASK [setup] *******************************************************************
ESTABLISH LOCAL CONNECTION FOR USER: copumpkin
other EXEC ( umask 22 && mkdir -p "$( echo $HOME/.ansible/tmp/ansible-tmp-1449199218.33-208217563604743 )" && echo "$( echo $HOME/.ansible/tmp/ansible-tmp-1449199218.33-208217563604743 )" )
other PUT /var/folders/f2/18n30h057611nz0cy1ng9bhc0000gn/T/tmpkVAINz TO /Users/copumpkin/.ansible/tmp/ansible-tmp-1449199218.33-208217563604743/setup
other EXEC LANG=en_US.UTF-8 LC_ALL=en_US.UTF-8 LC_MESSAGES=en_US.UTF-8 /usr/bin/python /Users/copumpkin/.ansible/tmp/ansible-tmp-1449199218.33-208217563604743/setup; rm -rf "/Users/copumpkin/.ansible/tmp/ansible-tmp-1449199218.33-208217563604743/" > /dev/null 2>&1
ok: [other]

TASK [debug msg=wow such local!] ***********************************************
task path: /Users/copumpkin/Sandbox/ansible-repro/test.yml:9
ok: [other] => {
    "changed": false, 
    "invocation": {
        "module_args": {
            "msg": "wow such local!"
        }, 
        "module_name": "debug"
    }, 
    "msg": "wow such local!"
}

PLAY RECAP *********************************************************************
localhost                  : ok=2    changed=0    unreachable=0    failed=0   
other                      : ok=2    changed
```

**Comment by jsmartin** [comment]
To me it seems like expected behavior.  Have you tried setting the ansible_python_interpreter for the "other" group to the same as localhost?

**Comment by copumpkin** [comment]
The issue is that in practice, `other` is a distinct machine from `localhost` and I don't want to override its python interpreter. But when I run tasks in the context of `other` with `connection: local`, I want to use the local python interpreter, not the remote one. Does that make sense or am I deeply confused? @bcoca seemed to think the behavior I was describing was reasonable on IRC, but I may just not have communicated very well.

To be super concrete:
- `localhost` is an odd Linux distro and doesn't even have a `/usr/bin/python` (in fact, it doesn't have a `/usr`). `ansible-playbook` is running against a python in a non-standard location and works fine in isolation.
- `other` is running a standard Linux distro with a valid `/usr/bin/python`. Ansible modules running remotely on `other` should use `/usr/bin/python`, so its `ansible_python_interpreter` should I think be `/usr/bin/python`.
- If I want to run occasional tasks against `other` but with `connection: local`, I can't use `other`'s `ansible_python_interpreter` because it doesn't even exist on `localhost`.

Does that help clarify the situation?

**Comment by jsmartin** [comment]
Ok, the piece I don't understand is, if `other` is a different machine, why would you be using `connection: local`?  Maybe you are thinking of doing a `local_action`on an `other` machine?

**Comment by copumpkin** [comment]
Mostly because I'm performing EC2 provisioning actions on it with a custom inventory script and the ansible EC2 modules. When I talked to bcoca on IRC, he suggested that `connection: local` was roughly equivalent to `delegate_to: localhost`, which was my mental picture of the thing, but this seems to indicate otherwise. Basically, I want the inventory variables relevant to the remote host, but I need to run some EC2 actions locally, without jumping into the other machine. Is there a better way to achieve that?

**Comment by bcoca** [comment]
might be related #13280

**Comment by copumpkin** [comment]
@bcoca thanks! Is there some sort of owner of this feature I should ping (or spec of it I could read) to better understand whether it's intentional behavior or not? I'm still trying to understand if I'm just doing something wrong (and if so, what I should be doing instead) or whether this is actually a bug (and ideally in that case if there's a simple workaround)

**Comment by bcoca** [comment]
it should use same python in both, we will be looking at this soon, but currently we are concentrating on the 2.0 release

**Comment by copumpkin** [comment]
@bcoca I see! I'll try to find a workaround in the meantime, but knowing that it's not PEBKAC helps me a lot, thanks.

**Comment by copumpkin** [comment]
For 1.9, I found a fairly simple workaround that re-injects the relevant variables into the executor. I'm not convinced it's the right place to put the change (and the location has changed in 2) but it does fix my test case to behave as I expect.

Inject into https://github.com/ansible/ansible/blob/v1.9.4-1/lib/ansible/runner/__init__.py#L690:

``` python
        elif host_connection == 'local':
            import re
            pattern = re.compile("^ansible_.*_interpreter$")
            for k, v in self.get_inject_vars("localhost").iteritems():
                if pattern.match(k):
                    inject[k] = v
            port = self.remote_port
```

and I end up with the right interpreter for the second task. If that seems like a reasonable approach to the problem, I can submit a simple PR.

**Comment by copumpkin** [comment]
Upon further investigation, the snippet above doesn't appear to be sufficient to fix all cases. It fixes my simple test case above, but a more complex one is still using the wrong python interpreter. I'll post another example later once I get time to reduce it.

**Comment by copumpkin** [comment]
Actually, I was wrong about being wrong. My patch snippet above does seem to fix the issue in all cases I was able to test.

@bcoca any feedback on whether this is the right approach?

It looks like @dagwieers is also encountering this problem.

**Comment by copumpkin** [comment]
An alternative would be to have `connection: local` magically desugar to `delegate_to: localhost`, but I don't know the source code well enough to be able to pull that off with confidence.

**Comment by bcoca** [comment]
that is exactly what it does, connection: local sets the delegated host to 'localhost'

**Comment by copumpkin** [comment]
@bcoca I get that that's how it's supposed to work, but it doesn't seem to behave the same way in practice, judging by @dagwieers's commit above and the test case I posted.

**Comment by copumpkin** [comment]
@bcoca what are the chances of getting this fixed in the first release of 2.0? I'm currently running a patched version (using the change I pasted above) locally on 1.9.\* but that isn't great.

**Comment by copumpkin** [comment]
If you can give me guidance, I'm not opposed to making my own PR, of course, but I'm having trouble even understanding your stance on the issue as I reported it, let alone whether my tiny "fix" is appropriate or not.

**Comment by bcoca** [comment]
last i tested, this is NOT an issue in 2.0 and we are we are not doing a 1.9.5 release until 2.0 is out, at that point we'll look at this in depth

**Comment by copumpkin** [comment]
@bcoca you tried my test case from the ticket above? I pasted my output for both 1.9 and [2.0](https://github.com/ansible/ansible/issues/13421#issuecomment-161857596) above and the output is identical and appears to still be broken in 2.0.

If I'm doing something wrong in that test (or my expectations are wrong), please tell me, but as far as I can tell this is not fixed in any 2.0 RC that I've seen. I even included the output against [devel](https://github.com/ansible/ansible/issues/13421#issuecomment-161862343) as of a month ago.

**Comment by bcoca** [comment]
_sigh_, it was fixed at one point but we seem to have regressed again ....

**Comment by bcoca** [comment]
this works correctly with `local_action` which does force the `delegate_to: localhost`, the case of overriding connection needs extra special casing for when `connection: local`

**Comment by joelthompson** [comment]
I just ran into this issue and would like to see it resolved in a sensible fashion, either by doing what @copumpkin suggests or providing something like a play-level delegate_to.

This particular use case was trying to run an Ansible playbook from a RHEL6 box with Python2.7 and Ansible installed inside a virtualenv. When running Ansible modules with connection: local, it breaks out of the vitualenv to use the system python interpreter (2.6), though I think it still maintained the virtualenv's python paths, so the system python2.6 interpeter tried to load the virtualenv's python 2.7 libraries, which didn't work.

**Comment by copumpkin** [comment]
@bcoca any updates? last you mentioned on IRC, there was some controversy as to how this should behave.

**Comment by bcoca** [comment]
So i believe #15024 should fix this and apply the correct vars

**Comment by bcoca** [comment]
closing via merged #15024
